monitoring/script.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package monitoring
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os/exec"
+
+	"github.com/gravitational/satellite/agent/health"
+	"github.com/gravitational/trace"
+)
+
+// NewScriptChecker returns a new script checker for the specified script.
+// dir can specify an alternative working directory for the script.
+// If dir is left unspecified, the running process's directory is used
+// as a working directory.
+func NewScriptChecker(script io.Reader, dir string, args ...string) health.Checker {
+	return scriptChecker{Reader: script, dir: dir, args: args}
+}
+
+// Name returns name of the checker.
+// Implements health.Checker
+func (r scriptChecker) Name() string {
+	return scriptCheckerID
+}
+
+func (r scriptChecker) Check(ctx context.Context, reporter health.Reporter) {
+	f, err := ioutil.TempFile("", "monitoring")
+	if err != nil {
+		reporter.Add(NewProbeFromErr(r.Name(), "failed to create script file",
+			trace.ConvertSystemError(err)))
+		return
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, r.Reader)
+	if err != nil {
+		reporter.Add(NewProbeFromErr(r.Name(), "failed to write script file",
+			trace.Wrap(err)))
+		return
+	}
+
+	args := append([]string{f.Name()}, r.args...)
+	cmd := exec.Command("bash", args...)
```

Comment:
This check is run in a user's cli right, so it should always have a functional path to locate bash on the system?

Reply:
Yes, hence I left it unspecified for `Command` to figure it out (which it will with [`LookPath`](https://golang.org/pkg/os/exec/#LookPath)) or did you mean anything else?

Reply:
No, I was just trying to make sure we don't envision any circumstances where this is run without an environment, like triggered through an rpc agent that is installed as a systemd unit.